src/models/Review.js

```diff
@@ -1,16 +1,38 @@
 const mongoose = require('mongoose');
+const Product = require('./Product');
 
-const reviewSchema = new mongoose.Schema({
-    rating: {
-        type: Number,
-        min: 0,
-        max: 5,
-        default: 1,
+const reviewSchema = new mongoose.Schema(
+    {
+        rating: {
+            type: Number,
+            min: 0,
+            max: 5,
+            default: 1,
+        },
+        comment: {
+            type: String,
+        },
     },
-    comment: {
-        type: String,
+    { timestamps: true, versionKey: false },
+);
+
+// pre-hook on remove : when review is deleted then it should be removed from product
+// and it's contribution to product's average rating should also be removed
+reviewSchema.pre(
+    'deleteOne',
+    { document: true, query: false },
+    async function (next) {
+        const product = await Product.findOne({ reviews: this._id });
+        // console.log(`product : \n ${product}`);
+        product.rating = (product.rating * product.reviews.length - this.rating)
+      / (product.reviews.length - 1);
+        product.reviews = product.reviews.filter(
```

Comment:
can we move the average logic calculation to a utils method, because our code should peel like onion, if someone is interested in knowing the actual logic than he can look at utils method, we make a method like
`Utils.getAverageRating()` that would be more clean

Reply:
https://mongoosejs.com/docs/guide.html#methods

Reply:
yes I'll make an instance method on the schema